pulser-core/pulser/hamiltonian_data/hamiltonian_data.py

```diff
@@ -0,0 +1,775 @@
+from dataclasses import replace
+from typing import Literal, cast
+
+import numpy as np
+from numpy.typing import ArrayLike
+from scipy.spatial.distance import cdist
+
+import pulser.math as pm
+from pulser.channels import Microwave, Raman, Rydberg
+from pulser.channels.base_channel import STATES_RANK, Channel, States
+from pulser.devices._device_datacls import COORD_PRECISION, BaseDevice
+from pulser.noise_model import NoiseModel, doppler_sigma
+from pulser.register import Register3D
+from pulser.register.base_register import BaseRegister, QubitId
+from pulser.sampler import sampler
+from pulser.sampler.samples import (
+    ChannelSamples,
+    SequenceSamples,
+    _PulseTargetSlot,
+)
+from pulser.sequence import Sequence
+
+from .noise_trajectory import NoiseTrajectory
+
+TRAP_WAVELENGTH = 0.85  # µm
+
+
+def _register_sigma_xy_z(
+    temperature: float, trap_waist: float, trap_depth: float
+) -> tuple[float, float]:
+    """Standard deviation for fluctuations in atom position in the trap.
+
+    - Plane fluctuation: 𝜎ˣʸ = √(T w²/(4 Uₜᵣₐₚ)), where T is temperature,
+      w is the trap waist and Uₜᵣₐₚ is the trap depth.
+    - Off plane fluctuation: 𝜎ᶻ = 𝜋 / 𝜆 √2 w 𝜎ˣʸ, where 𝜆 is the trap
+    wavelength with a constant value of 0.85 µm
+
+    Note: a k_B factor is absorbed in the trap depth (Uₜᵣₐₚ), so the units
+    of temperature and trap depth are the same.
+
+    Args:
+        temperature (float): Temperature (T) of the atoms in the trap
+            (in Kelvin).
+        trap_depth (float): Depth of the trap (Uₜᵣₐₚ)
+            (same units as temperature).
+        trap_waist (float): Waist of the trap (w) (in µmeters).
+
+    Returns:
+        tuple: The standard deviations of the spatial position fluctuations
+        in the xy-plane (register_sigma_xy) and along the z-axis
+        (register_sigma_z).
+    """
+    register_sigma_xy = math.sqrt(
+        temperature * trap_waist**2 / (4 * trap_depth)
+    )
+    register_sigma_z = (
+        math.pi
+        / TRAP_WAVELENGTH
+        * math.sqrt(2)
+        * trap_waist
+        * register_sigma_xy
+    )
+    return register_sigma_xy, register_sigma_z
+
+
+def _noisy_register(q_dict: dict, config: NoiseModel) -> Register3D:
+    """Add Gaussian noise to the positions of the register."""
+    register_sigma_xy, register_sigma_z = _register_sigma_xy_z(
+        config.temperature, config.trap_waist, cast(float, config.trap_depth)
+    )
+    atoms = list(q_dict.keys())
+    num_atoms = len(atoms)
+    positions = np.array(list(q_dict.values()))
+
+    if len(positions[0]) == 2:
+        positions = np.array(
+            [np.append(p, 0.0) for p in positions]
+        )  # Convert 2D positions to 3D
+
+    noise_xy = np.random.normal(0, register_sigma_xy, (num_atoms, 2))
+    noise_z = np.random.normal(0, register_sigma_z, num_atoms)
+    noise = np.column_stack((noise_xy, noise_z))
+    positions += noise
+    return Register3D({k: pos for (k, pos) in zip(atoms, positions)})
```

Comment:
I didn't think about this in Mauro's PR, but are we sure we want to cast everything to arrays here? It wouldn't be hard to preserve the gradient of the coordinates in the original `q_dict`

In any case, I would suggest adding a test where the register coordinates are torch tensors with gradient to see if there are no errors